### PR TITLE
feat: adds course enrollments endpoint for admin view

### PR DESCRIPTION
**Pull request by katrinan029**
## Description
Creates an endpoint for an admin of an enterprise to view all course enrollments for a learner in the admin portal.

https://2u-internal.atlassian.net/browse/ENT-10134

Screenshot 2025-03-06 at 4 58 41 PM

## Testing Instructions
1. follow this doc to add courses for a learner https://2u-internal.atlassian.net/wiki/spaces/SOL/pages/15141192/Creating+a+Course+Locally+in+Publisher+and+adding+to+an+Enterprise+Catalog#Prereqs.1
2. visit the endpoint http://localhost:18000/enterprise/api/v1/enterprise-course-enrollment-admin/?lms_user_id=<LMS_USER_ID>&enterprise_uuid=<ENTERPRISE_UUID> and confirm that you see the courses grouped by course status.

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [x] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [x] After versioned build finishes in [GitHub Actions](https://github.com/openedx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/openedx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - Trigger the '[Upgrade one Python dependency](https://github.com/openedx/edx-platform/actions/workflows/upgrade-one-python-dependency.yml)' action against master in edx-platform with new version number to generate version bump PR
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/api/v1/views/enterprise_course_enrollment.py [hunk 1]
```diff
@@ -86,3 +96,89 @@ def get_serializer_class(self):
...
+            return Response(
+                {"error": "Both 'lms_user_id' and 'enterprise_uuid' are required parameters."},
+                status=HTTP_400_BAD_REQUEST
+            )
+        enterprise_customer_user = get_object_or_404(
+            models.EnterpriseCustomerUser,
+            user_id=lms_user_id,
+            enterprise_customer__uuid=enterprise_uuid
+        )
+        enterprise_enrollments = models.EnterpriseCourseEnrollment.objects.filter(
```
Comment (jesperhodge) [inline enterprise/api/v1/views/enterprise_course_enrollment.py:142]: Considering separation of concerns:
We are in a view here, which is responsible for handling incoming requests and responding to them. The model `EnterpriseCourseEnrollment` is responsible for the relationship to its database table.
Might it be better to do the actual database queries in the model and just calling a function on the model, instead of operating on the queryset directly?
I always feel a bit uncomfortable when I see database queries in many different places than in the model, because it means that you may have different views duplicating the same case in slightly different ways, and it increases the chances of bugs. It would be nice to have a method `get_by_user_id` on the model that's well tested and any view just has to call that function instead of figuring out how to do database calls.

Comment (jesperhodge) [reply enterprise/api/v1/views/enterprise_course_enrollment.py:142]: It's also hard to see here if the code always works as expected. It would be great to have this as its own function with its own unit test that can consider any edge cases.

File: tests/test_enterprise/api/test_serializers.py [hunk 2]
```diff
@@ -628,3 +631,74 @@ def test_serialize_users(self):
...
+
+    @patch.object(HttpRequest, 'get_host', return_value='example.edx.org')
+    @patch('enterprise.models.CourseEnrollment')
+    @patch('enterprise.api.v1.serializers.get_certificate_for_user')
+    def test_serialization_basic(
+        self,
+        mock_get_certificate,
+        mock_course_enrollment_class,
+        _,
+    ):
```
Comment (jesperhodge) [inline tests/test_enterprise/api/test_serializers.py:657]: The serializer has a large number of conditionals, but I only see one dataset tested.
You may consider using the `@ddt` decorator to test this with different datasets.
Or break the serializer code into smaller functions tested on its own, which makes it easier to test edge cases.

**Review by jesperhodge — CHANGES_REQUESTED** [review]
Hey Katrina, nice work!
I haven't been able to test this yet locally. I'll probably go through the steps to add courses to a catalog locally too, but it will also suffice if you show me that this works on your machine.
I left some comments for improving test coverage and one refactor suggestion.

File: tests/test_enterprise/api/test_views.py [hunk 3]
```diff
@@ -9944,3 +9966,69 @@ def test_pagination_applied(self):
         self.assertIn("count", response.json())
         self.assertIn("next", response.json())
         self.assertIn("results", response.json())
+
+
+class EnterpriseCourseEnrollmentAdminViewSetTest(TestCase):
+    """Unit tests for EnterpriseCourseEnrollmentAdminViewSet."""
+    class MockSerializer:
```
[the remark this answers is not in the document]
Comment (jesperhodge) [reply tests/test_enterprise/api/test_views.py:9973]: Oh, got it.

File: enterprise/api/v1/views/enterprise_course_enrollment.py [hunk 4]
```diff
@@ -86,3 +96,89 @@ def get_serializer_class(self):
...
+    API views for the ``enterprise-course-enrollment-admin`` API endpoint.
+    """
+
+    queryset = models.EnterpriseCourseEnrollment.with_additional_fields.all()
+    serializer_class = serializers.EnterpriseCourseEnrollmentAdminViewSerializer
+    permission_classes = (permissions.IsAuthenticated, permissions.IsAdminUser)
+    pagination_class = EnterpriseCourseEnrollmentAdminPagination
+
+    @action(detail=False, methods=['get'])
+    def get_enterprise_course_enrollment(self, request):
```
Comment (jesperhodge) [inline enterprise/api/v1/views/enterprise_course_enrollment.py:121]: It's multiple enrollments, right? So maybe plural naming?

File: enterprise/api/v1/serializers.py [hunk 5]
```diff
@@ -430,6 +445,144 @@ class Meta:
...
+        representation['start_date'] = course_overview['start']
+        representation['end_date'] = course_overview['end']
+        representation['display_name'] = course_overview['display_name_with_default']
+        representation['org_name'] = course_overview['display_org_with_default']
+        representation['pacing'] = course_overview['pacing']
+        representation['is_revoked'] = instance.license.is_revoked if instance.license else False
+        representation['is_enrollment_active'] = instance.is_active
+        representation['mode'] = instance.mode
+
+        if CourseDetails:
```
Comment (jesperhodge) [inline enterprise/api/v1/serializers.py:489]: Could use a test, I'll give some feedback on how best to mock

File: enterprise/api/v1/serializers.py [hunk 6]
```diff
@@ -430,6 +445,144 @@ class Meta:
...
+        """
+        Get the appropriate course overview from the context.
+        """
+        for overview in self.context['course_overviews']:
+            if overview['id'] == course_run_id:
+                return overview
+
+        return None
+
+    def _get_exec_ed_course_run_status(self, course_details, certificate_info, enterprise_enrollment):
```
Comment (jesperhodge) [inline enterprise/api/v1/serializers.py:518]: test different conditions?

File: enterprise/api/v1/serializers.py [hunk 7]
```diff
@@ -430,6 +445,144 @@ class Meta:
...
+        has_started = datetime.now(pytz.utc) > start_date if start_date is not None else True
+        has_ended = datetime.now(pytz.utc) > end_date if end_date is not None else False
+
+        if has_ended or is_certificate_passing:
+            return CourseRunProgressStatuses.COMPLETED
+        if has_started:
+            return CourseRunProgressStatuses.IN_PROGRESS
+        return CourseRunProgressStatuses.UPCOMING
+
+    def _get_course_run_status(self, course_overview, certificate_info, enterprise_enrollment):
```
Comment (jesperhodge) [inline enterprise/api/v1/serializers.py:554]: test different conditions?

File: enterprise/api/v1/views/enterprise_course_enrollment.py [hunk 8]
```diff
@@ -86,3 +96,89 @@ def get_serializer_class(self):
...
+            return Response(
+                {"error": "Both 'lms_user_id' and 'enterprise_uuid' are required parameters."},
+                status=HTTP_400_BAD_REQUEST
+            )
+        enterprise_customer_user = get_object_or_404(
+            models.EnterpriseCustomerUser,
+            user_id=lms_user_id,
+            enterprise_customer__uuid=enterprise_uuid
+        )
+        enterprise_enrollments = models.EnterpriseCourseEnrollment.objects.filter(
```
Comment (jesperhodge) [inline enterprise/api/v1/views/enterprise_course_enrollment.py:142]: Could you test whether the enrollments are filtered correctly? Perhaps can be extracted into a function to test more easily

File: enterprise/api/v1/views/enterprise_course_enrollment.py [hunk 9]
```diff
@@ -86,3 +96,89 @@ def get_serializer_class(self):
...
+            filtered_enterprise_enrollments,
+            many=True,
+            context={
+                'request': request,
+                'enterprise_customer_user': enterprise_customer_user,
+                'course_overviews': course_overviews,
+            }
+        ).data
+        page = self.paginate_queryset(serialized_data)
+        if page is not None:
```
Comment (jesperhodge) [inline enterprise/api/v1/views/enterprise_course_enrollment.py:157]: Test pagination?

**Review by jesperhodge — APPROVED** [review]

**Comment by jesperhodge** [comment]
Great work on this!
